File: clients/jvm/src/main/kotlin/kivi/KiviClient.kt
// kivi JVM SDK (PLAN3 G3.3, K2+K3): Kotlin coroutines are the JVM async idiom
// (suspend functions + Flow); KiviBlockingClient is the Java-friendly sync
// facade over the same engine. The SDK constitution:
//   - an answer without a trace is unconstructable (TracedAnswer requires it);
//   - client-side verification is ON by default (hash+chain+numbering+seals);
//   - honest refusals are typed exceptions — NotFound is never a null.
package kivi

import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Metadata
import io.grpc.Status
import io.grpc.StatusException
import io.grpc.stub.MetadataUtils
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.runBlocking
import kivi.v1.KiviGrpcKt
import kivi.v1.KiviOuterClass as Pb

open class KiviException(message: String) : RuntimeException(message)
class NotFoundException(m: String) : KiviException(m)
class UnauthenticatedException(m: String) : KiviException(m)
class PreconditionException(m: String) : KiviException(m)
class InvalidArgumentException(m: String) : KiviException(m)
class UnavailableException(m: String) : KiviException(m)
class RateLimitedException(m: String) : KiviException(m)

data class Receipt(val no: Long, val offset: Long, val hash: String)

/** value(+raw json) + the events that established it + the derivation scope.
 * Constructing one without a trace throws — by design. */
data class TracedAnswer(val valueJson: String, val trace: List<Long>, val scope: Long) {
    init { require(trace.isNotEmpty()) { "an answer without a trace is unrepresentable" } }
}

data class VerifyReport(val ok: Boolean, val records: Long, val seals: Long,
                        val unsealedTail: Long, val tornTail: Boolean,
                        val hasDefect: Boolean, val defectNo: Long, val defectReason: String)

/** One traced similarity hit: the record's address, its score, and the raw
 * receipt itself — an untraced score does not exist. */
data class SimilarHit(val no: Long, val score: Float, val recordJson: String)

data class SimilarAnswer(val hits: List<SimilarHit>, val scope: Long, val model: String)

/** The result of a login: the bearer token plus its honest envelope. */
data class Session(val token: String, val role: String, val expiresUnix: Long)

/** One row of a paged view read: canonical key + that entry's JSON (trace inside). */
data class ViewEntry(val key: String, val valueJson: String)

/** One page of a keyset walk over a compiled view. nextKey == "" means done.
 * scope/hash stamp the SNAPSHOT: pass scope back as asOf and every later page
 * keeps describing that same moment. */
data class ViewPage(val entries: List<ViewEntry>, val nextKey: String,
                    val scope: Long, val hash: String)

private fun wrap(e: StatusException): KiviException = when (e.status.code) {
    Status.Code.NOT_FOUND -> NotFoundException(e.status.description ?: "not found")
    Status.Code.UNAUTHENTICATED -> UnauthenticatedException(e.status.description ?: "unauthenticated")
    Status.Code.FAILED_PRECONDITION -> PreconditionException(e.status.description ?: "precondition")
    Status.Code.INVALID_ARGUMENT -> InvalidArgumentException(e.status.description ?: "invalid")
    Status.Code.UNAVAILABLE -> UnavailableException(e.status.description ?: "unavailable")
    Status.Code.RESOURCE_EXHAUSTED -> RateLimitedException(e.status.description ?: "rate limited")
    else -> KiviException(e.status.toString())
}

class KiviClient(
    addr: String,
    token: String? = null,
    val verifyStreams: Boolean = true,
    channel: ManagedChannel? = null,
    private val ownsChannel: Boolean = channel == null,
) : AutoCloseable {
    private val chan = channel ?: ManagedChannelBuilder.forTarget(addr).usePlaintext().build()
    @Volatile private var stub: KiviGrpcKt.KiviCoroutineStub
    @Volatile private var codec: EntityCodec = EntityCodec.DEFAULT

    init {
        stub = stubWith(token)
    }

    private fun stubWith(token: String?): KiviGrpcKt.KiviCoroutineStub {
        var s = KiviGrpcKt.KiviCoroutineStub(chan)
        if (token != null) {
            val md = Metadata()
            md.put(Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER), "Bearer $token")
            s = s.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(md))
        }
        return s
    }

    /** Swap the bearer credential (after login or on rotation). */
    fun setToken(token: String?) { stub = stubWith(token) }

    /** Use a custom [EntityCodec] for the typed layer instead of kivi's
     *  built-in (dependency-free) mapper — e.g. a Jackson-backed one. Returns
     *  this client for fluent setup; the core never imports your library. */
    fun withCodec(codec: EntityCodec): KiviClient { this.codec = codec; return this }

    /** A view of this client over the SAME channel with ANOTHER identity —
     * per-caller credentials for stateless multi-tenant frontends. Closing a
     * view never closes the shared channel. */
    fun withBearer(token: String): KiviClient =
        KiviClient("", token, verifyStreams, chan, ownsChannel = false)

    override fun close() { if (ownsChannel) chan.shutdownNow() }

    private suspend fun <T> call(block: suspend () -> T): T =
        try { block() } catch (e: StatusException) { throw wrap(e) }

    /** Authenticate as a kivi USER and install the session token: from here
     * on every call runs with that user's role and every write is receipted
     * under their name. Security stays in ONE center — the server. */
    suspend fun login(username: String, password: String): Session = call {
        val r = stub.login(Pb.LoginRequest.newBuilder()
            .setUsername(username).setPassword(password).build())
        setToken(r.token)
        Session(r.token, r.role, r.expiresUnix)
    }

    /** The cheap orientation call: (headNo, headHash) — no audit runs. */
    suspend fun head(): Pair<Long, String> = call {
        stub.head(Pb.Empty.getDefaultInstance()).let { it.headNo to it.headHash }
    }

    /** Keyset pagination over a compiled view (table|graph|series). Pin a
     * consistent walk by passing page 1's scope as asOf. limit 0 = server
     * default (100, capped at 1000). */
    suspend fun viewPage(view: String, afterKey: String = "", limit: Long = 0,
                         asOf: Long? = null): ViewPage = call {
        val b = Pb.ViewPageRequest.newBuilder()
            .setView(view).setAfterKey(afterKey).setLimit(limit)
        if (asOf != null) b.setAsOf(asOf)
        val r = stub.queryViewPage(b.build())
        ViewPage(r.entriesList.map { ViewEntry(it.key, it.valueJson) },
            r.nextKey, r.scope, r.hash)
    }

    /** Traced semantic search: every hit is a record number + score + the
     * receipt itself; the answer names the model and the covered scope. */
    suspend fun similar(query: String, k: Long = 5): SimilarAnswer = call {
        val r = stub.similar(Pb.SimilarRequest.newBuilder().setQuery(query).setK(k).build())
        SimilarAnswer(r.hitsList.map { SimilarHit(it.no, it.score, it.recordJson) },
            r.scope, r.model)
    }

    suspend fun append(type: String, bodyJson: String): Receipt = call {
        stub.append(Pb.AppendRequest.newBuilder().setType(type).setBodyJson(bodyJson).build())
            .let { Receipt(it.no, it.offset, it.hash) }
    }

    /** Append a TYPED entity — pass a `@JvmRecord data class` (or Map/List);
     *  kivi serializes it with the active [EntityCodec] (default: kivi's own
     *  dependency-free mapper) and re-canonicalizes server-side, so field order
     *  never matters. "Entity in" — bring Jackson via [withCodec], or none. */
    suspend fun append(type: String, entity: Any): Receipt = append(type, codec.toJson(entity))

    /** Fetch record [no] and map its body into [cls] — the "entity out". The
     *  envelope is always parsed by kivi's core reader; only the body is handed
     *  to the active [EntityCodec]. */
    suspend fun <T> getRecordBodyAs(no: Long, cls: Class<T>, unseal: Boolean = false): T {
        @Suppress("UNCHECKED_CAST")
        val body = (Json.parse(getRecord(no, unseal)) as Map<String, Any?>)["body"]
        return codec.fromJson(Json.encode(body), cls)
    }

    suspend fun appendPrivate(type: String, bodyJson: String): Receipt = call {
        stub.appendPrivate(Pb.AppendRequest.newBuilder().setType(type).setBodyJson(bodyJson).build())
            .let { Receipt(it.no, it.offset, it.hash) }
    }

    /** Append a private (per-record encrypted) TYPED entity — an object, not a
     *  JSON string; serialized by the active [EntityCodec]. Erasable via [erase]. */
    suspend fun appendPrivate(type: String, entity: Any): Receipt = appendPrivate(type, codec.toJson(entity))

    suspend fun erase(no: Long, reason: String): Receipt = call {
        stub.erase(Pb.EraseRequest.newBuilder().setNo(no).setReason(reason).build())
            .let { Receipt(it.no, it.offset, it.hash) }
    }

    suspend fun seal(): Receipt = call {
        stub.seal(Pb.Empty.getDefaultInstance()).let { Receipt(it.no, it.offset, it.hash) }
    }

    suspend fun verifyLedger(): VerifyReport = call {
        stub.verify(Pb.Empty.getDefaultInstance()).let {
            VerifyReport(it.ok, it.records, it.seals, it.unsealedTail, it.tornTail,
                it.hasDefect, it.defectNo, it.defectReason)
        }
    }

    /** Traced read; asOf answers "what did we know when the ledger stopped
     * at record N" — time travel is free by design. */
    suspend fun table(subject: String, attribute: String, asOf: Long? = null): TracedAnswer = call {
        val b = Pb.TableRequest.newBuilder().setSubject(subject).setAttribute(attribute)
        if (asOf != null) b.setAsOf(asOf)
        stub.queryTable(b.build())
            .let { TracedAnswer(it.valueJson, it.traceList, it.scope) }
    }

    /** table()'s whole-row sibling (G15): every attribute known about one
     * subject, plus the union of every event that established one of them —
     * "what do we know about X?" without knowing which attributes to ask for
     * first. Same traced contract as table(). */
    suspend fun subject(subject: String, asOf: Long? = null): TracedAnswer = call {
        val b = Pb.SubjectRequest.newBuilder().setSubject(subject)
        if (asOf != null) b.setAsOf(asOf)
        stub.querySubject(b.build())
            .let { TracedAnswer(it.valueJson, it.traceList, it.scope) }
    }

    suspend fun why(trace: List<Long>): List<String> = call {
        stub.why(Pb.WhyRequest.newBuilder().addAllTrace(trace).build())
            .recordsList.map { it.recordJson }
    }

    suspend fun getRecord(no: Long, unseal: Boolean = false): String = call {
        stub.getRecord(Pb.GetRecordRequest.newBuilder().setNo(no).setUnseal(unseal).build())
            .recordJson
    }

    /** Verified replay: every record is re-hashed, chain+numbering checked and
     * seals Ed25519-verified CLIENT-SIDE (default). Raises VerificationException
     * on a lying stream. */
    fun replay(start: Long = 0, follow: Boolean = false,
               verify: Boolean? = null, checkSeals: Boolean = true): Flow<String> = flow {
        val checker = if (verify ?: verifyStreams) ChainChecker(checkSeals) else null
        try {
            stub.replay(Pb.ReplayRequest.newBuilder().setFrom(start).setFollow(follow).build())
                .collect { reply ->
                    checker?.check(reply.recordJson)
                    emit(reply.recordJson)
                }
        } catch (e: StatusException) {
            throw wrap(e)
        }
    }

    /** replay()'s type-filtered sibling (G14): the server drops non-matching
     * records before they cross the wire. Honesty note: because filtering
     * creates gaps by design, this does NOT carry replay()'s gapless/
     * chain-adjacency guarantee — each record's OWN hash/signature is still
     * verified, but numbering gaps and prev_hash discontinuities are
     * EXPECTED and never thrown as tampering. Use replay() when you must
     * prove no record was skipped. */
    fun subscribe(start: Long = 0, follow: Boolean = false, types: List<String> = emptyList(),
                  verify: Boolean? = null, checkSeals: Boolean = true): Flow<String> = flow {
        val checker = if (verify ?: verifyStreams) RecordIntegrityChecker(checkSeals) else null
        try {
            stub.subscribe(Pb.SubscribeRequest.newBuilder()
                    .setFrom(start).setFollow(follow).addAllTypes(types).build())
                .collect { reply ->
                    checker?.check(reply.recordJson)
                    emit(reply.recordJson)
                }
        } catch (e: StatusException) {
            throw wrap(e)
        }
    }
}

/** Java-friendly blocking facade over the coroutine client. */
class KiviBlockingClient(addr: String, token: String? = null, verify: Boolean = true) : AutoCloseable {
    private val inner = KiviClient(addr, token, verify)
    override fun close() = inner.close()

    fun append(type: String, bodyJson: String): Receipt = runBlocking { inner.append(type, bodyJson) }
    fun table(subject: String, attribute: String, asOf: Long? = null): TracedAnswer =
        runBlocking { inner.table(subject, attribute, asOf) }
    fun subject(subject: String, asOf: Long? = null): TracedAnswer =
        runBlocking { inner.subject(subject, asOf) }
    fun login(username: String, password: String): Session =
        runBlocking { inner.login(username, password) }
    fun head(): Pair<Long, String> = runBlocking { inner.head() }
    fun similar(query: String, k: Long = 5): SimilarAnswer =
        runBlocking { inner.similar(query, k) }
    fun viewPage(view: String, afterKey: String = "", limit: Long = 0, asOf: Long? = null): ViewPage =
        runBlocking { inner.viewPage(view, afterKey, limit, asOf) }
    fun verifyLedger(): VerifyReport = runBlocking { inner.verifyLedger() }
    fun why(trace: List<Long>): List<String> = runBlocking { inner.why(trace) }
    fun replayAll(verify: Boolean = true): List<String> = runBlocking {
        val out = ArrayList<String>()
        inner.replay(verify = verify).collect { out.add(it) }
        out
    }
    fun subscribeAll(types: List<String> = emptyList(), verify: Boolean = true): List<String> = runBlocking {
        val out = ArrayList<String>()
        inner.subscribe(types = types, verify = verify).collect { out.add(it) }
        out
    }
}

File: clients/dotnet/Kivi.Client/KiviClient.cs
// kivi .NET SDK (PLAN3 G3.4, K2+K3): async-first (Task — the .NET idiom) with
// a sync facade. The SDK constitution:
//   - an answer without a trace is unconstructable (TracedAnswer throws);
//   - client-side verification is ON by default (hash+chain+numbering+seals);
//   - honest refusals are typed exceptions — NotFound is never a null.
using System.Text.Json;
using Grpc.Core;
using Grpc.Net.Client;
using Kivi.V1;

namespace Kivi;

public class KiviException(string message) : Exception(message);
public class NotFoundException(string m) : KiviException(m);
public class UnauthenticatedException(string m) : KiviException(m);
public class PreconditionException(string m) : KiviException(m);
public class InvalidArgumentException(string m) : KiviException(m);
public class UnavailableException(string m) : KiviException(m);
public class RateLimitedException(string m) : KiviException(m);

public readonly record struct Receipt(long No, long Offset, string Hash);

/// <summary>valueJson + the events that established it + the derivation scope.
/// Constructing one without a trace throws — by design.</summary>
public sealed record TracedAnswer
{
    public string ValueJson { get; }
    public IReadOnlyList<long> Trace { get; }
    public long Scope { get; }

    public TracedAnswer(string valueJson, IReadOnlyList<long> trace, long scope)
    {
        if (trace.Count == 0)
            throw new ArgumentException("an answer without a trace is unrepresentable");
        (ValueJson, Trace, Scope) = (valueJson, trace, scope);
    }
}

public readonly record struct LedgerReport(bool Ok, long Records, long Seals,
    long UnsealedTail, bool TornTail, bool HasDefect, long DefectNo, string DefectReason);

/// <summary>One traced similarity hit: the record's address, its score, and
/// the raw receipt itself — an untraced score does not exist.</summary>
public readonly record struct SimilarHit(long No, float Score, string RecordJson);

public readonly record struct SimilarAnswer(IReadOnlyList<SimilarHit> Hits,
    long Scope, string Model);

/// <summary>The result of a login: the bearer token plus its honest envelope.</summary>
public readonly record struct Session(string Token, string Role, long ExpiresUnix);

/// <summary>One row of a paged view read: canonical key + that entry's JSON (trace inside).</summary>
public readonly record struct PagedEntry(string Key, string ValueJson);

/// <summary>One page of a keyset walk over a compiled view. NextKey == "" means
/// done. Scope/Hash stamp the SNAPSHOT: pass Scope back as asOf and every later
/// page keeps describing that same moment.</summary>
public readonly record struct PagedView(IReadOnlyList<PagedEntry> Entries,
    string NextKey, long Scope, string Hash);

/// <summary>One directed, traced relation: Source —Relation→ Target, set by
/// record No.</summary>
public readonly record struct GraphEdge(string Relation, string Target, long No);

/// <summary>A node found by a traversal, with its depth and the edge record
/// numbers along the shortest discovering path (the proof).</summary>
public readonly record struct GraphReached(string Node, long Depth, IReadOnlyList<long> Trace);

/// <summary>One step of a path: From —Relation→ To, via edge record No.</summary>
public readonly record struct GraphHop(string From, string Relation, string To, long No);

public sealed class KiviClient : IDisposable
{
    private readonly GrpcChannel _channel;
    private readonly bool _ownsChannel;
    private readonly V1.Kivi.KiviClient _stub;
    private Metadata _md = [];
    public bool VerifyStreams { get; }

    public KiviClient(string addr, string? token = null, bool verify = true)
    {
        _channel = GrpcChannel.ForAddress(addr.StartsWith("http") ? addr : $"http://{addr}");
        _ownsChannel = true;
        _stub = new V1.Kivi.KiviClient(_channel);
        if (token != null) _md.Add("authorization", $"Bearer {token}");
        VerifyStreams = verify;
    }

    private KiviClient(GrpcChannel shared, string? token, bool verify)
    {
        _channel = shared;
        _ownsChannel = false;
        _stub = new V1.Kivi.KiviClient(_channel);
        if (token != null) _md.Add("authorization", $"Bearer {token}");
        VerifyStreams = verify;
    }

    /// <summary>A view of this client over the SAME channel with ANOTHER
    /// identity — per-caller credentials for stateless multi-tenant
    /// frontends. Disposing a view never closes the shared channel.</summary>
    public KiviClient WithBearer(string token) => new(_channel, token, VerifyStreams);

    /// <summary>Swap the bearer credential (after login or on rotation).</summary>
    public void SetToken(string? token)
    {
        var md = new Metadata();
        if (token != null) md.Add("authorization", $"Bearer {token}");
        _md = md;
    }

    public void Dispose() { if (_ownsChannel) _channel.Dispose(); }

    private static KiviException Wrap(RpcException e) => e.StatusCode switch
    {
        StatusCode.NotFound => new NotFoundException(e.Status.Detail),
        StatusCode.Unauthenticated => new UnauthenticatedException(e.Status.Detail),
        StatusCode.FailedPrecondition => new PreconditionException(e.Status.Detail),
        StatusCode.InvalidArgument => new InvalidArgumentException(e.Status.Detail),
        StatusCode.Unavailable => new UnavailableException(e.Status.Detail),
        StatusCode.ResourceExhausted => new RateLimitedException(e.Status.Detail),
        _ => new KiviException(e.Status.ToString()),
    };

    private async Task<T> Call<T>(Func<Task<T>> f)
    {
        try { return await f().ConfigureAwait(false); }
        catch (RpcException e) { throw Wrap(e); }
    }

    public Task<Receipt> AppendAsync(string type, string bodyJson) => Call(async () =>
    {
        var r = await _stub.AppendAsync(new AppendRequest { Type = type, BodyJson = bodyJson }, _md);
        return new Receipt(r.No, r.Offset, r.Hash);
    });

    public Task<Receipt> AppendPrivateAsync(string type, string bodyJson) => Call(async () =>
    {
        var r = await _stub.AppendPrivateAsync(new AppendRequest { Type = type, BodyJson = bodyJson }, _md);
        return new Receipt(r.No, r.Offset, r.Hash);
    });

    // -- typed "entity in, entity out" (System.Text.Json, already in the runtime) --

    /// <summary>Snake_case field names — the kivi ecosystem convention
    /// (policy_id, premium_cents). PascalCase C# records map to it both ways.</summary>
    internal static readonly JsonSerializerOptions JsonOpts =
        new() { PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower };

    /// <summary>Append a TYPED entity — you pass an object, kivi serializes it
    /// (and re-canonicalizes server-side, so field order never matters).</summary>
    public Task<Receipt> AppendAsync<T>(string type, T entity) =>
        AppendAsync(type, JsonSerializer.Serialize(entity, JsonOpts));

    /// <summary>Append a private (per-record encrypted) TYPED entity — an object,
    /// not a JSON string. Erasable later; see EraseAsync.</summary>
    public Task<Receipt> AppendPrivateAsync<T>(string type, T entity) =>
        AppendPrivateAsync(type, JsonSerializer.Serialize(entity, JsonOpts));

    /// <summary>Decode an already-fetched record's BODY into a TYPED entity —
    /// the "entity out" for records from ReplayAsync/SubscribeAll/Why/GetRecord.</summary>
    public static T BodyAs<T>(string recordJson)
    {
        using var doc = JsonDocument.Parse(recordJson);
        return doc.RootElement.GetProperty("body").Deserialize<T>(JsonOpts)!;
    }

    /// <summary>Fetch record <c>no</c> and decode its body into a typed entity.</summary>
    public Task<T> GetRecordBodyAsAsync<T>(long no, bool unseal = false) => Call(async () =>
        BodyAs<T>(await GetRecordAsync(no, unseal)));

    public Task<Receipt> EraseAsync(long no, string reason) => Call(async () =>
    {
        var r = await _stub.EraseAsync(new EraseRequest { No = no, Reason = reason }, _md);
        return new Receipt(r.No, r.Offset, r.Hash);
    });

    public Task<Receipt> SealAsync() => Call(async () =>
    {
        var r = await _stub.SealAsync(new Empty(), _md);
        return new Receipt(r.No, r.Offset, r.Hash);
    });

    public Task<LedgerReport> VerifyLedgerAsync() => Call(async () =>
    {
        var r = await _stub.VerifyAsync(new Empty(), _md);
        return new LedgerReport(r.Ok, r.Records, r.Seals, r.UnsealedTail,
            r.TornTail, r.HasDefect, r.DefectNo, r.DefectReason);
    });

    /// <summary>Traced read; asOf answers "what did we know when the ledger
    /// stopped at record N" — time travel is free by design.</summary>
    public Task<TracedAnswer> TableAsync(string subject, string attribute,
        long? asOf = null) => Call(async () =>
    {
        var req = new TableRequest { Subject = subject, Attribute = attribute };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.QueryTableAsync(req, _md);
        return new TracedAnswer(r.ValueJson, r.Trace.ToList(), r.Scope);
    });

    /// <summary>TableAsync's whole-row sibling (G15): every attribute known
    /// about one subject, plus the union of every event that established one
    /// of them — "what do we know about X?" without knowing which attributes
    /// to ask for first. Same traced contract as TableAsync.</summary>
    public Task<TracedAnswer> SubjectAsync(string subject, long? asOf = null) => Call(async () =>
    {
        var req = new SubjectRequest { Subject = subject };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.QuerySubjectAsync(req, _md);
        return new TracedAnswer(r.ValueJson, r.Trace.ToList(), r.Scope);
    });

    /// <summary>Authenticate as a kivi USER and install the session token:
    /// every later call runs with that user's role and every write is
    /// receipted under their name. Security stays in ONE center — the server.</summary>
    public Task<Session> LoginAsync(string username, string password) => Call(async () =>
    {
        var r = await _stub.LoginAsync(new LoginRequest
        { Username = username, Password = password }, _md);
        SetToken(r.Token);
        return new Session(r.Token, r.Role, r.ExpiresUnix);
    });

    /// <summary>The cheap orientation call: (HeadNo, HeadHash) — no audit runs.</summary>
    public Task<(long HeadNo, string HeadHash)> HeadAsync() => Call(async () =>
    {
        var r = await _stub.HeadAsync(new Empty(), _md);
        return (r.HeadNo, r.HeadHash);
    });

    /// <summary>Keyset pagination over a compiled view (table|graph|series).
    /// Pin a consistent walk by passing page 1's Scope as asOf; limit 0 =
    /// server default (100, capped at 1000).</summary>
    public Task<PagedView> ViewPageAsync(string view, string afterKey = "",
        long limit = 0, long? asOf = null) => Call(async () =>
    {
        var req = new ViewPageRequest { View = view, AfterKey = afterKey, Limit = limit };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.QueryViewPageAsync(req, _md);
        return new PagedView(
            r.Entries.Select(e => new PagedEntry(e.Key, e.ValueJson)).ToList(),
            r.NextKey, r.Scope, r.Hash);
    });

    /// <summary>Direct outgoing relation edges of <paramref name="node"/>. asOf
    /// pins the graph to a record number (null = current).</summary>
    public Task<(IReadOnlyList<GraphEdge> Edges, bool Truncated)> GraphNeighborsAsync(
        string node, long? asOf = null) => Call(async () =>
    {
        var req = new GraphNeighborsRequest { Node = node };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.GraphNeighborsAsync(req, _md);
        return ((IReadOnlyList<GraphEdge>)r.Edges
            .Select(e => new GraphEdge(e.Relation, e.Target, e.No)).ToList(), r.Truncated);
    });

    /// <summary>Every node reachable from <paramref name="node"/> within
    /// <paramref name="depth"/> hops (0 = server default), each with its edge
    /// trace. limit 0 = server default.</summary>
    public Task<(IReadOnlyList<GraphReached> Nodes, bool Truncated)> GraphReachableAsync(
        string node, long depth = 0, long limit = 0, long? asOf = null) => Call(async () =>
    {
        var req = new GraphReachableRequest { Node = node, Depth = depth, Limit = limit };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.GraphReachableAsync(req, _md);
        return ((IReadOnlyList<GraphReached>)r.Nodes
            .Select(x => new GraphReached(x.Node, x.Depth, x.Trace.ToList())).ToList(), r.Truncated);
    });

    /// <summary>Shortest edge path from <paramref name="from"/> to
    /// <paramref name="to"/> within <paramref name="depth"/> hops (0 = server
    /// default). Found=false when none exists inside the bound.</summary>
    public Task<(IReadOnlyList<GraphHop> Hops, bool Found, bool Truncated)> GraphPathAsync(
        string from, string to, long depth = 0, long? asOf = null) => Call(async () =>
    {
        var req = new GraphPathRequest { From = from, To = to, Depth = depth };
        if (asOf is long n) req.AsOf = n;
        var r = await _stub.GraphPathAsync(req, _md);
        return ((IReadOnlyList<GraphHop>)r.Hops
            .Select(h => new GraphHop(h.From, h.Relation, h.To, h.No)).ToList(), r.Found, r.Truncated);
    });

    /// <summary>Traced semantic search: every hit is a record number + score
    /// + the receipt itself; the answer names the model and covered scope.</summary>
    public Task<SimilarAnswer> SimilarAsync(string query, long k = 5) => Call(async () =>
    {
        var r = await _stub.SimilarAsync(new SimilarRequest { Query = query, K = k }, _md);
        return new SimilarAnswer(
            r.Hits.Select(h => new SimilarHit(h.No, h.Score, h.RecordJson)).ToList(),
            r.Scope, r.Model);
    });

    public Task<List<string>> WhyAsync(IEnumerable<long> trace) => Call(async () =>
    {
        var req = new WhyRequest();
        req.Trace.AddRange(trace);
        var r = await _stub.WhyAsync(req, _md);
        return r.Records.Select(x => x.RecordJson).ToList();
    });

    public Task<string> GetRecordAsync(long no, bool unseal = false) => Call(async () =>
    {
        var r = await _stub.GetRecordAsync(new GetRecordRequest { No = no, Unseal = unseal }, _md);
        return r.RecordJson;
    });

    /// <summary>Verified replay: every record is re-hashed, chain+numbering
    /// checked and seals Ed25519-verified CLIENT-SIDE (default). Throws
    /// VerificationException on a lying stream.</summary>
    public async IAsyncEnumerable<string> ReplayAsync(long start = 0, bool follow = false,
        bool? verify = null, bool checkSeals = true)
    {
        var checker = (verify ?? VerifyStreams) ? new ChainChecker(checkSeals) : null;
        using var call = _stub.Replay(new ReplayRequest { From = start, Follow = follow }, _md);
        while (true)
        {
            bool ok;
            try { ok = await call.ResponseStream.MoveNext(default).ConfigureAwait(false); }
            catch (RpcException e) { throw Wrap(e); }
            if (!ok) yield break;
            var json = call.ResponseStream.Current.RecordJson;
            checker?.Check(json);
            yield return json;
        }
    }

    /// <summary>ReplayAsync's type-filtered sibling (G14): the server drops
    /// non-matching records before they cross the wire, so a consumer that
    /// only cares about one or two event types never receives (and
    /// discards) everything else.
    ///
    /// Honesty note: because the server may drop records, this does NOT
    /// carry ReplayAsync's gapless/chain-adjacency guarantee. With
    /// verification on (default), each delivered record's OWN hash is
    /// recomputed (and its Ed25519 signature checked, for kivi.seal
    /// records) — proving "byte-exact what the ledger holds" — but
    /// numbering gaps and prev_hash discontinuities are EXPECTED and never
    /// thrown as tampering. A consumer that must prove it received every
    /// record uses ReplayAsync instead.</summary>
    public async IAsyncEnumerable<string> SubscribeAsync(long start = 0, bool follow = false,
        IEnumerable<string>? types = null, bool? verify = null, bool checkSeals = true)
    {
        var checker = (verify ?? VerifyStreams) ? new RecordIntegrityChecker(checkSeals) : null;
        var req = new SubscribeRequest { From = start, Follow = follow };
        if (types is not null) req.Types_.AddRange(types);
        using var call = _stub.Subscribe(req, _md);
        while (true)
        {
            bool ok;
            try { ok = await call.ResponseStream.MoveNext(default).ConfigureAwait(false); }
            catch (RpcException e) { throw Wrap(e); }
            if (!ok) yield break;
            var json = call.ResponseStream.Current.RecordJson;
            checker?.Check(json);
            yield return json;
        }
    }

    // -- sync facade (documented blocking wrappers over the async engine) ----

    public Receipt Append(string type, string bodyJson) =>
        AppendAsync(type, bodyJson).GetAwaiter().GetResult();
    public Receipt Append<T>(string type, T entity) =>
        AppendAsync(type, entity).GetAwaiter().GetResult();
    public Receipt AppendPrivate<T>(string type, T entity) =>
        AppendPrivateAsync(type, entity).GetAwaiter().GetResult();
    public T GetRecordBodyAs<T>(long no, bool unseal = false) =>
        GetRecordBodyAsAsync<T>(no, unseal).GetAwaiter().GetResult();
    public TracedAnswer Table(string subject, string attribute, long? asOf = null) =>
        TableAsync(subject, attribute, asOf).GetAwaiter().GetResult();
    public TracedAnswer Subject(string subject, long? asOf = null) =>
        SubjectAsync(subject, asOf).GetAwaiter().GetResult();
    public Session Login(string username, string password) =>
        LoginAsync(username, password).GetAwaiter().GetResult();
    public (long HeadNo, string HeadHash) Head() => HeadAsync().GetAwaiter().GetResult();
    public SimilarAnswer Similar(string query, long k = 5) =>
        SimilarAsync(query, k).GetAwaiter().GetResult();
    public PagedView ViewPage(string view, string afterKey = "", long limit = 0, long? asOf = null) =>
        ViewPageAsync(view, afterKey, limit, asOf).GetAwaiter().GetResult();
    public LedgerReport VerifyLedger() => VerifyLedgerAsync().GetAwaiter().GetResult();
    public List<string> Why(IEnumerable<long> trace) => WhyAsync(trace).GetAwaiter().GetResult();
    public List<string> ReplayAll(bool verify = true)
    {
        var outp = new List<string>();
        var e = ReplayAsync(verify: verify).GetAsyncEnumerator();
        try { while (e.MoveNextAsync().AsTask().GetAwaiter().GetResult()) outp.Add(e.Current); }
        finally { e.DisposeAsync().AsTask().GetAwaiter().GetResult(); }
        return outp;
    }
    public List<string> SubscribeAll(IEnumerable<string>? types = null, bool verify = true)
    {
        var outp = new List<string>();
        var e = SubscribeAsync(types: types, verify: verify).GetAsyncEnumerator();
        try { while (e.MoveNextAsync().AsTask().GetAwaiter().GetResult()) outp.Add(e.Current); }
        finally { e.DisposeAsync().AsTask().GetAwaiter().GetResult(); }
        return outp;
    }
}
